src/target/nrf91.c

```diff
@@ -81,25 +134,180 @@ static void nrf91_add_flash(target_s *target, uint32_t addr, size_t length, size
 	flash->erased = 0xff;
 	target_add_flash(target, flash);
+
+	/* add separate UICR flash */
+	target_flash_s *flash_uicr = calloc(1, sizeof(*flash_uicr));
+	if (!flash_uicr) { /* calloc failed: heap exhaustion */
+		DEBUG_WARN("calloc: failed in %s\n", __func__);
+		return;
+	}
+
+	flash_uicr->start = 0xff8000U;
+	flash_uicr->length = 0x1000U;
+	flash_uicr->blocksize = 0x1000U;
+	flash_uicr->erase = nrf91_uicr_flash_erase;
+	flash_uicr->write = nrf91_flash_write;
+	flash_uicr->erased = 0xff;
+	target_add_flash(target, flash_uicr);
+
+}
+
+static bool nrf91_mass_erase(target_s *target)
+{
+	adiv5_access_port_s *ap = cortex_ap(target);
+	adiv5_access_port_s ctrl_ap = {
+		.dp = ap->dp,
+		.apsel = 0x4U,
+	};
+
+	return nrf91_ctrl_ap_mass_erase(&ctrl_ap);
 }
 
 bool nrf91_probe(target_s *target)
 {
 	adiv5_access_port_s *ap = cortex_ap(target);
 
-	if (ap->dp->version < 2U)
+	if (ap->dp->version < 2U || ap->dp->target_partno != 0x90U)
 		return false;
 
-	switch (ap->dp->target_partno) {
-	case 0x90:
-		target->driver = "Nordic nRF9160";
-		target->target_options |= TOPT_INHIBIT_NRST;
-		target_add_ram(target, 0x20000000, 256U * 1024U);
-		nrf91_add_flash(target, 0, 4096U * 256U, 4096U);
-		break;
-	default:
+	uint32_t buf[3] = {0};
+
+	ap->dp->mem_read(ap, buf, 0x00FF0140, 3*sizeof(uint32_t));
+
+	gdb_outf("nRF%04x %4s%4s detected!\n", buf[0], (const char *)&buf[2], (const char *)&buf[1]);
+
+	switch (ap->dp->target_revision) {
+		case 0:
+		case 1:
+		case 2:
+			target->driver = "Nordic nRF9160";
+			break;
+		case 3:
+			target->driver = "Nordic nRF91x1";
+			break;
+		default:
+			target->driver = "Nordic nRF91";
+	}
+
+	target->target_options |= TOPT_INHIBIT_NRST;
+	target_add_ram(target, 0x20000000, 256U * 1024U);
+	nrf91_add_flash(target, 0, 4096U * 256U, 4096U);
+
+	target->mass_erase = nrf91_mass_erase;
+
+
+	/* Persist AP access if uninitialized (only needed for devices with hardenend APPROTECT) */
+	/* TODO: can't add this, it breaks normal loading */
+//	if (ap->dp->target_revision > 2) {
+//		/* read UICR.APPROTECT */
+//		ap->dp->mem_read(ap, &buf[0], UICR_APPROTECT, sizeof(uint32_t));
+//		/* read UICR.SECUREAPPROTECT */
+//		ap->dp->mem_read(ap, &buf[1], UICR_SECUREAPPROTECT, sizeof(uint32_t));
+//
+//		target_mem_write32(target, NRF91_NVMC_CONFIG, NRF91_NVMC_CONFIG_WEN);
+//
+//		while (target_mem_read32(target, NRF91_NVMC_READY) == 0) {
+//			platform_delay(1);
+//			DEBUG_INFO("Waiting for NVMC to become ready\n");
+//		}
+//
+//		if (target_mem_read32(target, UICR_APPROTECT) == UICR_ERASED_VAL) {
+//			target_mem_write32(target, UICR_APPROTECT, UICR_APPROTECT_UNPROTECT_VAL);
+//		}
+//		if (target_mem_read32(target, UICR_SECUREAPPROTECT) == UICR_ERASED_VAL) {
+//			target_mem_write32(target, UICR_SECUREAPPROTECT, UICR_APPROTECT_UNPROTECT_VAL);
+//		}
+//
+//		target_mem_write32(target, NRF91_NVMC_CONFIG, NRF91_NVMC_CONFIG_REN);
+//	}
+
+	return true;
+}
+
+
+
+
+static bool nrf91_rescue_do_recover(target_s *t)
```

Comment:
Please avoid single letter variable naming for this, please use `target` instead. Likewise for `nrf91_rescue_probe()` on return from `target_new()`.

This is the style we've been gradually moving the whole code base to.